Replace clobber parameter with overwrite

evertrol opened the pull request:
`overwrite` can now be used instead of `clobber`. The `clobber`
parameter is deprecated, and an AstropyDeprecationWarning will be
issued if it is used instead of `overwrite`. If both happen to be
used, a ValueError is raised.

evertrol commented:
A note on the implementation:
I've implemented this at the `file.py` level, in the `_File` class, which is used by most output routines. Exceptions are `diff.py` and `table.py`, which use their own output routines.

A note on the unit tests:
I've replaced all ``clobber`` occurrences with ``overwrite``. Except for the newly written unit tests, there is thus no test to see if ``clobber`` actually works with the old routines.
The new unit tests are perhaps a bit awkwardly placed: I wanted to make use of the ``FitsTestCase`` temporary file utilities, but the actual tests of ``overwrite`` vs ``clobber`` seem better placed outside any test class (but that would require a different mechanism for temporary files).

MSeifert04 commented:
I'm not really sure if it may be a problem that you added the ``overwrite`` parameter in between existing parameters. Personally I would have preffered something along the lines of ``func(..., overwrite=None, ..., **kwargs)``. Placing the ``overwrite`` at the position where the old ``clobber`` was to prevent problems with code using clobber (or subsequent parameters) as positional parameters. Especially for `HDUList.writeto()` I often used it as positional parameter.

Also catching ``clobber`` as ``kwarg`` will hide it from the auto-completion of runtime environments (like Notebooks) and makes it much easier to remove it some day without having to alter every function definition again. You could propagate it through the internals and check for ``clobber`` in the end.

On the other hand I'm not really sure if using ``**kwargs`` just to deprecate one argument is such a good alternative because one would need to check if any unsuitable keyword argument or typo-parameter was catched as well.

Thank you for this thorough PR. I'll have a closer look in the next days.

MSeifert04 commented on astropy/io/fits/convenience.py:
```diff
@@ -409,8 +409,16 @@ def writeto(filename, data, header=None, output_verify='exception',
+        If `True`, overwrite the output file if exists.
+        The use of clobber is deprecated; use overwrite instead.
+
+    overwrite : bool, optional
```
Sphinx has a [`versionadded`](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded) and [`deprecated`](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-deprecated)-directive. I haven't seen it regularly used but I think it could make sense here.

For `clobber` this would be something like 

```
.. deprecated:: 1.3
   Use the ``overwrite`` argument instead.
```

and for overwrite: `.. versionadded:: 1.3`

MSeifert04 commented on astropy/io/fits/diff.py:
```diff
@@ -149,21 +150,45 @@ def report(self, fileobj=None, indent=0, clobber=False):
         return_string = False
         filepath = None
 
+        if isinstance(clobber, bool) and isinstance(overwrite, bool):
```
It seems like `clobber` just had to be "truthy" but not necessarily be a boolean. 

Say for example I set `clobber=1, overwrite=False` then this wouldn't throw the error even though I specified both of them.

evertrol replied:
Correct. It is a bit of a trade-off between the user doing the right thing (using booleans) and checking for every possibility (e.g., ``clobber=0`` equates to ``clobber=False``, not ``clobber=None``).
I felt that if someone uses ``clobber=1, overwrite=False``, it's up to them to live with the result. 
In cases where the value for clobber or overwrite derives from an evaluation, it'll generally be a bool, so such non-explicit values should give the correct result.

pllim replied:
I don't like this particular implementation. `io.fits` is so widely used and people are so used to `clobber` that such a transition should be as painless to them as possible. I posted some pseudo-code in my other comment.

MSeifert04 commented on astropy/io/fits/hdu/table.py:
```diff
@@ -999,14 +1009,27 @@ def dump(self, datafile=None, cdfile=None, hfile=None, clobber=False):
         plain text (ASCII) files.
         """
 
+        if isinstance(clobber, bool) and isinstance(overwrite, bool):
```
This seems identical to the code used in `report`. Maybe you could factor the code this code into a seperate function?

evertrol replied:
Do you mean identical to the lines in `diff.py` and `file.py`?
I'm not sure whether it's worth creating a small function in e.g. `util.py`, or to have some code duplication. The three occurrences of this code reflect the three different output directions that exist in ``astropy.io.fits``.

evertrol commented:
@MSeifert04 Thanks for catching that; I had totally overlooked the fact that keyword arguments might be used as positional arguments in existing code. I had simply placed ``overwrite`` next to ``clobber``.
I'll wait a bit to see what other people might say, but otherwise I'll move ``overwrite`` to the far right in the function definitions. (I'm in favour of being explicit, i.e., not using ``**kwargs``.)
Once we get to astropy version 2, we can move it back to a more suitable place, if so wanted.

I'll have a look at the Sphinx directives.

saimn commented:
+1 for adding overwrite, but I think that deprecating clobber is very annoying for a very little gain. clobber must be used in a lot of code, and I fear people will get warnings for a long time before having a clobber-free code base. This is the kind of things that will bother people. So they will filter warnings, and all other `AstropyDeprecationWarning` by the way, making these warnings useless.
Also supporting older Astropy versions (like the 1.0) will require to add checks for Astropy's version to avoid the warnings.
Supporting both option is not so complicated.

pllim commented on CHANGES.rst:
```diff
@@ -78,6 +78,10 @@ API Changes
 
+  - FITS writers now accept an ``overwrite`` argument. If ``clobber``
+    is used, an ``AstropyDeprecationWarning`` is issued. If both are
+    used, a ``ValueError`` is raised.
```
Need to append PR number here. See other entries for example.

pllim commented on astropy/io/fits/convenience.py:
```diff
@@ -383,7 +383,7 @@ def delval(filename, keyword, *args, **kwargs):
 
 def writeto(filename, data, header=None, output_verify='exception',
-            clobber=False, checksum=False):
+            clobber=None, overwrite=None, checksum=False):
```
Shouldn't the default `overwrite` be `False`?

saimn replied:
It is used to know which of the two options is explicitly set.

evertrol replied:
Neither, since otherwise it is not possible to test for inconsistency, when somehow both clobber and overwrite are used (but with inconsistent values).

pllim commented:
Replacing `clobber` with `overwrite` actually makes the unified I/O more consistent. This has been discussed elsewhere before although not linked to this PR. I agree with @MSeifert04 that it is better to "hide" clobber in `**kwargs`. Maybe something like the pseudo-code below:

```python
if clobber in kwargs:
    issue_deprecation_warning
    overwrite = clobber
```

Or if you want to be stricter, maybe something like (but I think the former is nicer to people who already use `clobber`):

```python
if clobber in kwargs:
    issue_deprecation_warning
    if overwrite is not None:
        raise ValueError('You have to explicitly set overwrite=None to use clobber')
    overwrite = clobber
```

pllim commented on astropy/io/fits/convenience.py:
```diff
@@ -409,8 +409,16 @@ def writeto(filename, data, header=None, output_verify='exception',
     clobber : bool, optional
-        If `True`, and if filename already exists, it will overwrite
-        the file.  Default is `False`.
+        If `True`, overwrite the output file if exists.
```
If we move `clobber` into `kwargs`, this text can go into `.. note::` section in the docstring. The code should also issue an explicit DeprecationWarning with this text.

pllim commented on astropy/io/fits/convenience.py:
```diff
@@ -420,8 +428,8 @@ def writeto(filename, data, header=None, output_verify='exception',
         hdu = PrimaryHDU(data, header=header)
-    hdu.writeto(filename, clobber=clobber, output_verify=output_verify,
-                checksum=checksum)
+    hdu.writeto(filename, clobber=clobber, overwrite=overwrite,
```
Should only pass in `overwrite` here. It is too confusing to pass in both. You can do `overwrite = clobber` prior to this call.

evertrol replied:
That won't work, since all handling of `clobber` versus `overwrite` is done in the `_File` class. And `hdu.writeto` implicitly calls that, so parameters need to be passed through.

pllim replied:
I wonder if we can apply the "move `clobber` to `kwargs`" action item here...

pllim commented on astropy/io/fits/convenience.py:
```diff
@@ -689,7 +697,7 @@ def info(filename, output=None, **kwargs):
 
 def tabledump(filename, datafile=None, cdfile=None, hfile=None, ext=1,
-              clobber=False):
+              clobber=None, overwrite=None):
```
See API comments above.

pllim commented on astropy/io/fits/convenience.py:
```diff
@@ -718,7 +726,16 @@ def tabledump(filename, datafile=None, cdfile=None, hfile=None, ext=1,
     clobber : bool
-        Overwrite the output files if they exist.
+        If `True`, overwrite the output file if exists.
+        The use of clobber is deprecated; use overwrite instead.
```
See doc comments above.

pllim commented on astropy/io/fits/convenience.py:
```diff
@@ -744,7 +761,7 @@ def tabledump(filename, datafile=None, cdfile=None, hfile=None, ext=1,
 
         # Dump the data from the HDU to the files
-        f[ext].dump(datafile, cdfile, hfile, clobber)
+        f[ext].dump(datafile, cdfile, hfile, clobber, overwrite)
```
*Replace* `clobber` with `overwrite` (see above).

pllim commented on astropy/io/fits/convenience.py:
```diff
@@ -755,8 +772,8 @@ def tabledump(filename, datafile=None, cdfile=None, hfile=None, ext=1,
 def tdump(filename, datafile=None, cdfile=None, hfile=None, ext=1,
-          clobber=False):
-    tabledump(filename, datafile, cdfile, hfile, ext, clobber)
+          clobber=None, overwrite=None):
```
*Replace* `clobber` with `overwrite` (see above).

pllim commented on astropy/io/fits/diff.py:
```diff
@@ -131,9 +133,8 @@ def identical(self):
-    def report(self, fileobj=None, indent=0, clobber=False):
-        """
-        Generates a text report on the differences (if any) between two
+    def report(self, fileobj=None, indent=0, clobber=None, overwrite=None):
```
See API comment above.

pllim commented on astropy/io/fits/diff.py:
```diff
@@ -149,21 +150,45 @@ def report(self, fileobj=None, indent=0, clobber=False):
-            fileobj is specified as a path.
+            If ``True`` and fileobj is a path, overwrite the output
+            file if it exist.
+            The use of clobber is deprecated; use overwrite instead.
```
See doc comment above.

pllim commented on astropy/io/fits/hdu/table.py:
```diff
@@ -999,14 +1009,27 @@ def dump(self, datafile=None, cdfile=None, hfile=None, clobber=False):
+            if clobber is None:
+                overwrite = False  # Be explicit
+            else:
+                warnings.warn(
```
I only see the warning being explicitly issued here. Is this sufficient?

pllim commented:
> no test to see if `clobber` actually works with the old routines

We should at least keep one test that uses the old `clobber` keyword successfully and make sure it still overwrites an existing file, just because it is so widely used.

evertrol commented:
I see a few different opinions and suggestions above, and I'm not sure how now to proceed.

saimn makes the correct argument that `clobber` in a (py)fits related context is widely used, and deprecating (and finally removing) the keyword may cause problems for users further down the road (as such a change will take a while to propagate through dependencies).
I'm in favour of removing it eventually, and then better start issuing warnings early. I don't expect `clobber` to be removed before astropy version 2.
Users that still use dependencies that haven't replaced `clobber` by `overwrite` may then be forced to use astropy 1.x though, while they may want to use some astropy 2.x features in other parts of their code.

There is also the suggestion of moving `clobber` into `kwargs`. That does create a problem for cases where a positional argument is used to set `clobber`, unless of course `overwrite` is put at that place in the function arguments.

I'm not really sure how much checking there should be to see whether both `overwrite` and `clobber` are used, and if they are inconsistent. It does pose the problem that the default value has to be `None` instead of `False` in the function signature, since otherwise it's not possible to verify that only one of the keywords has been used. But perhaps it should be allowed to use both (with a DeprecationWarning), and just check for inconsistency in that case? (But this raises immedate ValueErrors with an `overwrite=False` default while a user uses `clobber=True`.)
Related to that: as per MSeifert04's note, should we support anything truthy or falsely, or should be enforce strict boolean values? My preference is the latter.

pllim commented:
I propose:

1. Move `clobber` to `kwargs` and put `overwrite` in its place, as you suggested, to support positional argument call signature.
2. Never set `clobber` nor `overwrite` to `None` (IMHO, that just complicates the logic). If `clobber` is present, always just clobber but issue a `DeprecationWarning` and ignore `overwrite`. This is to satisfy backward compatibility until it is removed, without raising any error.

Feel free to discuss further. Those are just my personal thoughts.

MSeifert04 commented:
I agree with @pllim but I want to propose another point:

- Put the logic how to resolve the clobber/overwrite parameters in a seperate function. (https://github.com/astropy/astropy/pull/5171#discussion_r70358482) That way it is garantueed that each high/low level functionality behaves _exactly the same_. It makes it much easier to edit, move or expand it if the need arises.

bsipocz commented:
@pllim - I think the 2nd part of your point 2 is not backward compatibility, but a possibility to cause a big headache when the actual removal of ``clobber`` happens. 

Currently one gets a big ``TypeError`` when specifying both clobber and overwrite, so I'm 👎 on changing that to a ``DeprecationWarning`` while accepting it. If anyone takes the effort to change their legacy code to have ``overwrite``, they may remove clobber right away.

So I would suggest to raise the ``DeprecationWarning`` whenever ``clobber`` is used, but keep throwing errors on the user if they try to use both.

MSeifert04 commented:
Another way but be using a decorator to catch the parameters passed to the functions (this would work best at the high-level-functions) and resolve the conflicts before any function is actually executed, something (very roughly) like this:

```
import functools

def deprecateArgument(argumentname, position):
    def outerwrapper(function):
        @functools.wraps(function)
        def wrapper(*args, **kwargs):
            # The only way to have clobber inside the function is
            # that it is passed as kwarg because the clobber was
            # renamed to overwrite
            if 'clobber' in kwargs:
                clobber = kwargs.pop('clobber')
                print('clobber is deprecated.')
                
                # Check if an overwrite was given
                overwrite_in_args = len(args) > position
                overwrite_in_kwargs = 'overwrite' in kwargs
                
                # Annoy the user if he annoys us. They shouldn't have
                # specified overwrite and clobber!
                if overwrite_in_args or overwrite_in_kwargs:
                    raise TypeError('cannot specify overwrite and clobber.')
                # Pass in the clobber.
                else:
                    kwargs['overwrite'] = clobber
                
            return function(*args, **kwargs)
        return wrapper
    return outerwrapper
```

pllim commented:
@bsipocz , whatever is fine. I'll let the main package maintainer decide which idea to adopt.

MSeifert04 commented:
@pllim ATM, I think there is no maintainer for `io.fits`: https://github.com/astropy/astropy/wiki/Sub-package-maintainers. Or am I wrong?

evertrol commented:
I've now implemented the approach suggested by @MSeifert04 (thanks!), using a decorator. I made the decorator slightly more generic: it takes two arguments, the old and new keyword. See the [diff for astropy.io.fits.util in this PR](https://github.com/astropy/astropy/pull/5171/files#diff-26cfeae4c79ad73cf80f51a96a899063R989) for the current implementation.
That makes me wonder if it's worth writing a more generic decorator for replacing keywords in functions & methods, and add it to [astropy.utils.decorators](https://github.com/astropy/astropy/blob/master/astropy/utils/decorators.py) instead. That module already contains decorators to mark a function, class or attribute as deprecated (though in this case, it's more a keyword replacement).

Documentation-wise, I've left `clobber` in the parameter section of the doc-string, with the suggested `..deprecated` directive (and `..versionadded` for `overwrite`). I haven't build the docs, so I don't know what it looks like. Perhaps the deprecation note is better placed at the end of the doc-string (outside the parameter list), and `clobber` completely removed from the parameter documentation (since the function definition doesn't contain it either).
Using `..versionchanged` at the end is another option (with no `clobber` in the parameter list, and no `..versionadded` to the `overwrite` description):
```python
.. versionchanged: 1.3
   overwrite replaces the clobber keyword. Clobber is deprecated.
```
An example in the Python or numpy documentation would be good to use, but I haven't found one for this particular situation.

evertrol commented:
Note on the travis failure: there seems to be a bug introduced in scipy. Using scipy 0.17, the tests run fine. With 0.18, about half fail (notably those where the returned value from `scipy.stats.circmean()` is less than pi. I suspect a modulo change, possibly an intentional behaviour change). 
Looks like [this SciPy commit](https://github.com/scipy/scipy/commit/6f799a807bde2b4391857136fd3cbf87321710f1) introduced that change.

pllim commented on astropy/io/fits/diff.py:
```diff
@@ -26,8 +27,10 @@
 from ...extern.six.moves import zip, xrange, reduce
 from ...utils import indent
 from ...utils.compat.funcsigs import signature
+from ...utils.exceptions import AstropyDeprecationWarning
```
Remove unused import.

pllim commented on astropy/io/fits/file.py:
```diff
@@ -19,11 +19,11 @@
 from ...utils.decorators import classproperty
-from ...utils.exceptions import AstropyUserWarning
-
+from ...utils.exceptions import AstropyUserWarning, AstropyDeprecationWarning
```
Remove unused import.

pllim commented on astropy/io/fits/hdu/table.py:
```diff
@@ -26,13 +26,13 @@
 from ....utils import deprecated, lazyproperty
 from ....utils.compat import suppress
-from ....utils.exceptions import AstropyUserWarning
+from ....utils.exceptions import AstropyUserWarning, AstropyDeprecationWarning
```
Remove unused import.

pllim commented on astropy/io/fits/util.py:
```diff
@@ -982,3 +984,74 @@ def _get_array_mmap(array):
+
+def replaced_argument(oldkeyword, newkeyword):
+    """Decorate a function to replace 'oldkeyword' by 'newkeyword, but
+    allow for (deprecated) use of 'oldkeyword''
```
I remember vaguely that a double single quotes is a way to escape a single quote. I would recommend you build the documentation locally and check that everything compiles properly and looks like it should be.

evertrol replied:
That's a typo, combined with me not liking double backticks; but I guess Sphinx likes them, so I'll change that to double backticks.

pllim commented on astropy/io/fits/util.py:
```diff
@@ -982,3 +984,74 @@ def _get_array_mmap(array):
+    return posargs
+
+
+def replaced_argument(oldkeyword, newkeyword):
```
I agree that this is better suited to be placed in `astropy.utils.decorators`, not here. And indeed it looks a lot cleaner using a decorator like this! :+1:

MSeifert04 commented on astropy/io/fits/util.py:
```diff
@@ -982,3 +984,74 @@ def _get_array_mmap(array):
+            posargs = get_posargs(function)
+            try:
+                position = posargs.index(newkeyword)
+            except ValueError:
```
Wouldn't it be better to let the ValueError rise? Not being able to find the new keyword is definitly a misuse of the decorator.

evertrol replied:
My initial thought behind it was that this allows for a clearer error message to the user. But if this ValueError occurs, then it's a mistake on the developer side (the function signature and decorator keywords don't add up), so yes, better to let it rise.

pllim commented on astropy/io/fits/tests/test_util.py:
```diff
@@ -26,3 +27,20 @@ def test():
         pytest.raises(KeyboardInterrupt, test)
+
+
+def test_deprecated():
```
If the decorator moves to `astropy.utils.decorators`, this test also has to move with it.

MSeifert04 commented on CHANGES.rst:
```diff
@@ -78,6 +78,10 @@ API Changes
 
+  - FITS writers now accept an ``overwrite`` argument. If ``clobber``
+    is used, an ``AstropyDeprecationWarning`` is issued. If both are
+    used, a ``ValueError`` is raised. [#5171]
```
It's a TypeError now.

MSeifert04 commented:
I would be slightly in favor of removing `clobber` completly from the docstrings given that it is effectivly removed from the function signature. Your suggested `versionchanged` looks good to me. But wrap the parameters in double backticks then (and lowercase clobber in the second sentence).

The decorator looks good to me :+1:

MSeifert04 commented on astropy/io/fits/util.py:
```diff
@@ -982,3 +984,74 @@ def _get_array_mmap(array):
+        def wrapper(*args, **kwargs):
+            # Determine the position of the possible newkeyword
+            # positional parameter
+            posargs = get_posargs(function)
```
You could do this in a more lazy fashion. I don't know how fast/slow ``inspect`` is but it might be faster to do something like this:

```
def decorator(function):
    posarg = None
    @functools.wraps(function)
    def wrapper(*args, **kwargs):
        nonlocal posarg
        if posarg is None:
            posarg = random.randint(0, 100) # put your logic here
        ...
    return wrapper
```

MSeifert04 commented on astropy/io/fits/util.py:
```diff
@@ -982,3 +984,74 @@ def _get_array_mmap(array):
+
+    """
+    try:
+        posargs = list(inspect.signature(function).parameters.keys())
```
Please use `astropy.utils.compat.funcsigs.signature` instead of this try/except.

MSeifert04 commented on astropy/io/fits/tests/test_hdulist.py:
```diff
@@ -782,3 +782,31 @@ def test_extname_in_hdulist(self):
+            assert (str(warning_lines[0].message) == "The use of the "
+                    "clobber parameter is deprecated. Please use "
+                    "overwrite in the future.")
+        # All four illegal combinations
```
I don't think we should check these cases (those 4 illegal combinations). They are not part of the `io.fits` functions anymore and if (or as soon as) we remove clobber completly we will have to remove these anyway.

MSeifert04 commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -2417,6 +2418,42 @@ def test_reference_leak(self, tmpdir):
+                assert (str(warning_lines[0].message) == "The use of the "
+                        "clobber parameter is deprecated. Please use "
+                        "overwrite in the future.")
+            # All four illegal combinations
```
same here: These 4 cases should be tested for the decorator.

MSeifert04 commented on astropy/io/fits/tests/test_diff.py:
```diff
@@ -633,14 +636,44 @@ def test_file_output_clobber_safety(self):
+                    "clobber parameter is deprecated. Please use "
+                    "overwrite in the future.")
+        # All four illegal combinations
+        with pytest.raises(TypeError) as exc:
```
same here

evertrol commented:
I think it's worth by now to put in a separate PR for the new decorator, and then update this PR accordingly? Otherwise this would be a PR with two rather disparate changes.

pllim commented:
@evertrol , I don't think it is unusual to add a new decorator in `astropy.utils.decorators` as part of this PR. It is much more book-keeping work to add this in `fits.utils` and then move it, because then you will have to deprecate your new decorator.

MSeifert04 commented:
@evertrol Could you explain what you mean by "seperate PR"? Do you mean a follow-up PR which moves the decorator to the utils and modifies it? Or do you mean removing the decorator completly from this PR and implement it in another PR?

@pllim We need to deprecate features that haven't been released?

FWIW I had some time and modified the decorator to some extend and if requested I could open another PR for it. But I don't think we should put this PR on hold until any decorator is merged.

pllim commented:
@MSeifert04 , not sure. But if people start to use the one on dev, they will be very confused why decorator suddenly disappear. Also, there is no guarantee that it will be moved before a release. Why go through an extra hoop when we don't have to?

Also, you can open a PR against @evertrol 's branch for this PR if you want to get fancy. :smile:

evertrol commented:
I mean a new PR for this more generic decorator, as it has little directly to with overwrite/clobber: by now, it has become a decorator for making it easier to transition when a function replaces a keyword.
Obviously, the current decorator will be moved out of this PR, and this PR will be updated just import it from `astropy.utils.decorators` (once the other PR is accepted & merged).

It would not be so much a follow-up PR, more a prerequisite for this PR.

MSeifert04 commented:
@evertrol The decorator required for this PR is now merged into master. You'd need to rebase your PR to fix the conflicts and make use of the decorator.

Sorry for the long delay.

evertrol commented:
I've updated this PR to use the new ``deprecated_renamed_argument`` decorator (#5171).

I've also taken the liberty to change the last remaining ``clobber`` parameter to ``overwrite``, in the ``vo`` subpackage (and added a tiny bit of unit test coverage to remove the `#pragma no cover` I came across there).

pllim commented:
I am okay with the VO stuff. @MSeifert04 , you want to have another look at this before merge?

evertrol commented on astropy/vo/client/tests/test_vos_catalog.py:
```diff
@@ -186,6 +186,9 @@ def test_write(self):
         db.to_json(outfile)
+        with pytest.raises(OSError) as exc:
+            db.to_json(outfile)
+        assert str(exc.value).endswith("test_1.json exists.")
```
I went with the style given by the [pytest](http://doc.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions) documentation, which puts the assertion outside the with block.
In fact, most (if not all) of the astropy test code follows this style.

MSeifert04 replied:
@pllim This is the only not-adressed comment left, should it be changed?

pllim replied:
I guess if tests pass, then it is fine as-is.

evertrol commented:
There is one issue I noticed just now when building the docs: the `.. versionadded::` directive doesn't always get properly rendered. For example, if it follows a parameter listing, it gets rendered as just another parameter (including the leading two periods). An example of that is `HDUList.writeto`.
It looks like this is an issue with numpydoc, which follows somewhat strict guidelines with [the section order](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections), causing directives to be ignored when inside a `Parameters` subsection. Would be good if someone could confirm this issue, and even better if someone knows a way around this.

MSeifert04 commented:
@evertrol Good that you mention that. I've included a PR that should fix this #5492. If you have time could you check if that PR adresses all the malformed sphinx docstrings? I've checked with `HDUList.writeto` and it looks reasonably well.

evertrol commented:
@MSeifert04 Looks good. The only place where it still fails is for `Header.totextfile`, but that is because the doc string does not follow standard conventions. 
I'll add a commit to this PR that adds a summary line to `totextfile` (and `fromtextfile` above it), unless people veto that (as it has nothing to do with the clobber -> overwrite replacement).

pllim commented:
> a commit to this PR that adds a summary line

Since code freeze is coming up, I have no problem with fixing minor docs stuff in this PR.

So, does #5492 has to be merged *before* this one?

MSeifert04 commented:
@pllim I'd say we wait for 1-2 days until we merge this PR. In case #5492 should be merged (controversial, see discussion there) it should be merged at least roughly at the same time as this PR otherwise we vandalize some documentation in between. In case #5492 rather removes the "versionadded" parts, then this PR should include the manually inserted "versionadded" directives.

In both cases the PRs should be merged contemporaneous but in no particular order. 😄

MSeifert04 commented on astropy/io/fits/convenience.py:
```diff
@@ -408,9 +410,11 @@ def writeto(filename, data, header=None, output_verify='exception',
-        If `True`, and if filename already exists, it will overwrite
-        the file.  Default is `False`.
+    overwrite : bool, optional
+        If ``True``, overwrite the output file if exists.
```
not exactly sure but the `,` seems weird. Also it should note that if `False` and the file exists and OSError is raised.

MSeifert04 replied:
this also should mention that the default can be changed in `io.fits.Conf`.

evertrol replied:
About the `OSError`: shouldn't this be under a separate header `Raises`? Or would that be considered when the function itself raises the error (here the functionality is just passed on,  all the way up to `astropy.io.fits.file`)?

The actual exception raised differs, since [Python 3.3 merged `IOError`](https://docs.python.org/3/whatsnew/3.3.html#pep-3151-reworking-the-os-and-io-exception-hierarchy) and a few other exceptions into `OSError`, and `astropy.io.file._overwrite_existing` explicitly raises an `IOError`.
Perhaps this should be changed to raising an `OSError` (which for people using 2.7 and catching that error, makes that a backward incompatible change).
Otherwise, the docstring could be amended to include the 2.7 case: "raises an ``OSError`` (``IOError`` on Python 2) if the file exists and ``overwrite`` is ``False``."
(And arguably, Python 3 could raise `FileExistsError` instead.)

I don't understand the remark about `io.fits.Conf`: I don't see anything related to overwriting files there.
I would think it awkward design as well, for a keyword with explicit `True` or `False` values (having a default of `None` could trigger such default behaviour, but the default is `False` everywhere afaict).
I'm probably misreading this.

evertrol replied:
Should the default value for a parameter (here ``False``) be mentioned in the parameter list, and if so, where (if there is a preference)? Directly following the type and ", optional" line, at the start of the description, or at the end?

The default value is of course glanced immediately from the function definition, but so is the fact that the keyword is optional, and the latter is still explicitly listed in the parameter list.

MSeifert04 replied:
Sorry, I confused the conf with `mmap`, my bad. Then please also mention that the default is ``False``.

MSeifert04 replied:
I like the suggestion about "raises an OSError (IOError on Python 2) if the file exists and overwrite is False." but I don't like the ``Raises`` section if it's completly dependant on the _one_ parameter.

evertrol replied:
On the comma ;-) 
I read it as "If it is true, then ...", requiring a comma; in particular because there are two present tense verbs in that sentence. ([Additional reasoning](http://english.stackexchange.com/questions/52091/comma-after-introductory-phrases).)

MSeifert04 commented on astropy/io/fits/hdu/base.py:
```diff
@@ -332,10 +333,10 @@ def readfrom(cls, fileobj, checksum=False, ignore_missing_end=False,
                 checksum=False):
-        """
-        Write the HDU to a new file.  This is a convenience method to
+        """Write the HDU to a new file.  This is a convenience method to
```
See the other comment about the docstring start.

MSeifert04 commented on astropy/io/fits/hdu/base.py:
```diff
@@ -1565,7 +1567,8 @@ def match_header(cls, header):
+    def writeto(self, name, output_verify='exception', overwrite=False,
                 checksum=False):
         """
         Works similarly to the normal writeto(), but prepends a default
```
Missing ``versionchanged`` directive.

MSeifert04 replied:
is adressed below.

MSeifert04 commented on astropy/io/fits/hdu/table.py:
```diff
@@ -27,13 +27,13 @@
 from ..fitsrec import FITS_rec, _get_recarray_field, _has_unicode_fields
 from ..header import Header, _pad_length
 from ..util import _is_int, _str_to_num
-
```
Any reason why you removed that blank line. I think this should seperate the `fits`-imports from the general astropy imports.

MSeifert04 commented on astropy/io/fits/header.py:
```diff
@@ -654,14 +656,18 @@ def tofile(self, fileobj, sep='', endcard=True, padding=True,
+
+            .. versionchanged:: 1.3
+               ``overwrite`` replaces the deprecated ``clobber`` argument
+
```
No objection but generally the docstring doesn't end with a blank line.

MSeifert04 commented on astropy/io/fits/header.py:
```diff
@@ -687,7 +693,8 @@ def tofile(self, fileobj, sep='', endcard=True, padding=True,
     @classmethod
     def fromtextfile(cls, fileobj, endcard=False):
-        """
+        """Read a header from a simple text file or file-like object
```
Missing a `.` at the end of the sentence. I would also include a link to `Header.fromfile` here.

MSeifert04 commented on astropy/io/fits/header.py:
```diff
@@ -696,16 +703,18 @@ def fromtextfile(cls, fileobj, endcard=False):
-        """
+    @deprecated_renamed_argument('clobber', 'overwrite', '1.3')
+    def totextfile(self, fileobj, endcard=False, overwrite=False):
+        """Write the header as text to a file or a file-like object
```
See comment to `fromtxtfile`.

MSeifert04 commented on astropy/io/fits/hdu/base.py:
```diff
@@ -352,18 +353,19 @@ def writeto(self, name, output_verify='exception', clobber=False,
         checksum : bool
             When `True` adds both ``DATASUM`` and ``CHECKSUM`` cards
             to the header of the HDU when written to the file.
+
```
Any reason to include a blank line here?

MSeifert04 commented on astropy/io/fits/header.py:
```diff
@@ -696,16 +703,18 @@ def fromtextfile(cls, fileobj, endcard=False):
+    @deprecated_renamed_argument('clobber', 'overwrite', '1.3')
+    def totextfile(self, fileobj, endcard=False, overwrite=False):
+        """Write the header as text to a file or a file-like object
+
```
Missing `versionchanged`-directive.

evertrol replied:
I've left it out for this case, since there is no parameter list either. Will add it to the overall docstring.

MSeifert04 replied:
It's an alias for another function so it makes sense not to have the parameter list copied here.

👍 for putting it in the overall docstring.

MSeifert04 commented on astropy/io/fits/convenience.py:
```diff
@@ -408,9 +410,11 @@ def writeto(filename, data, header=None, output_verify='exception',
+        If ``True``, overwrite the output file if exists.
+
+        .. versionchanged:: 1.3
+           ``overwrite`` replaces the deprecated ``clobber`` argument
```
Please include the dot for each versionchanged directive.

MSeifert04 commented:
So I've left some more stylistic comments. Overall this looks great 👍

evertrol commented on astropy/io/fits/diff.py:
```diff
@@ -134,9 +135,9 @@ def identical(self):
-        Generates a text report on the differences (if any) between two
+    @deprecated_renamed_argument('clobber', 'overwrite', '1.3')
+    def report(self, fileobj=None, indent=0, overwrite=False):
+        """Generates a text report on the differences (if any) between two
```
As far as I'm aware, it doesn't change the overall rendering (both through Sphinx or when using the help in the interpreter); I just went with PEP257. (The same goes for blank lines at the end of a docstring, which often come about from emacs' fill-paragraph). I've changed it to the astropy convention.

MSeifert04 replied:
It's not so much about conventions, it's more that everyone has a different preference and if everyone would change the style this could well lead to some unintended edit-wars with a lot of unnecessary commit-conflicts.

MSeifert04 commented on CHANGES.rst:
```diff
@@ -176,6 +176,10 @@ API Changes
 
+  - FITS writers now accept an ``overwrite`` argument. If ``clobber``
+    is used, an ``AstropyDeprecationWarning`` is issued. If both are
+    used, a ``TypeError`` is raised. [#5171]
```
Could you list the changed functions here?

MSeifert04 replied:
is adressed below.

astrofrog replied:
I would rephrase this to say that 'The ``clobber`` argument in FITS writers has been renamed to ``overwrite``' (then leave the bit about deprecation).

evertrol commented:
Another (more general) question on the style of docstrings: the astropy docs state "Variable, module and class names should be written between single back-ticks". But how is this for function parameters (`` `overwrite` ``  or ` ``overwrite`` `?) and constants such as `True`, `False` and `None`? (I presume the single backticks are for Sphinx to provide role links, which would not really work for function parameters and constants.)
This seems inconsistent throughout astropy, though that's not a problem, as long as there a standard out there that things can converge to.

(Of course, these are all minute details that nobody should really care about, but with the detailed list of comments provided by @MSeifert04 , I'd now like to know.)

MSeifert04 commented:
single backticks indicate that it must be rendered as a link by numpydoc/sphinx. This is possible for everything that has a documentation (functions, classes, methods, modules). 
It's possible to link to `True`, `None`, etc. but it's not possible to link to parameters. Therefore you **must** enclose parameters with double backticks (or face a failing sphinx build).

I don't think there is any standard conventions but I'd say linking to numpy/astropy objects is a good idea while linking to standard python types (str, int, float) should be avoided. Note that to too many links and code-like blocks make the documentation hard to read, so it's also hard to define a hard limit here.

Note that this is just my opinion and the maintainers of astropy might have different preferences.

evertrol commented:
The Travis failure appears to be to a set of (otherwise unrelated) warnings that causes Sphinx to complain, and a possibly stalled build.

MSeifert04 commented:
Not all sphinx warnings are flukes: 

```
['Header.totextfile(fileobj, endcard=False, overwrite=False)', ':module: astropy.io.fits', '', 'Write the header as text to a file or a file-like object.', '', 'Equivalent to::', '', "    >>> Header.tofile(fileobj, sep='\\n', endcard=False,", '    ...               padding=False, overwrite=overwrite)', '', '.. versionchanged:: 1.3', '   ``overwrite`` replaces the deprecated ``clobber`` argument.', '', '.. seealso::', '', '    :func:`tofile`', '', '    :obj:`..`', '        ``overwrite`` replaces the deprecated ``clobber`` argument.']:17: WARNING: py:obj reference target not found: .
```

MSeifert04 commented on astropy/io/fits/header.py:
```diff
@@ -688,24 +695,40 @@ def tofile(self, fileobj, sep='', endcard=True, padding=True,
+
+        See Also
+        --------
+        :func:`tofile`.
```
I think `tofile` without `:func` and without the backticks is sufficient.

evertrol replied:
That is unexpected behaviour to me, that a "See Also" section turns everything into a link. Which causes the period to become problematic (I did wonder where the period went when building the docs locally).

evertrol replied:
Perhaps, until the point where one likes to add a tiny comment about that link and its relevancy. I guess I favour explicit (links) over implicit.

MSeifert04 replied:
It's possible to add a comment, just use:

`tofile : some really important and short description`

MSeifert04 replied:
See also (pun intended!) : https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#id5 the ninth point shows this.

MSeifert04 commented on astropy/io/fits/header.py:
```diff
@@ -688,24 +695,40 @@ def tofile(self, fileobj, sep='', endcard=True, padding=True,
+
+        See Also
+        --------
+        :func:`fromfile`.
```
Here as well.

evertrol commented:
With regards to the Sphinx warning, fixing it appears to be not that easy somewhat: I am now stuck with a "See Also" section which includes the text from the `..versionchanged` directive in its green box (this is specifically for `io.fits.Header.totextfile`), *as well as* having the text just before the section (which is where it is located in the `totextfile` docstring).
Removing the directive from the `totextfile` docstring makes the text before "See Also" go away (as expected), but the "overwrite replaces the deprecated clobber argument" remains inside the green box.
Removing the entire build directory or using `setup.py build_docs -a -E -l` doesn't make a difference, which I find really confusing.
Something is probably getting cached, but I can't find where.

MSeifert04 commented:
That's weird. I can confirm this.

Maybe just remove the directive from to and fromtextfile and file an issue at astropy-helpers.

MSeifert04 commented:
Oh, I see ... the problem is that #5492 isn't merged yet and the directive is added to the end of the function as well as your manual directive. 😅

evertrol commented:
@MSeifert04 Correct. I did that yesterday just fine by rebasing this PR on top of #5492, but I clearly woke up on the wrong side of the bed or too fresh today, so I forgot about it. That's why I didn't come across any Sphinx errors locally yesterday (even if the provided links were a bit overdone). 
I didn't see any errors for the most recently pushed update either (when locally build and rebased on #5492), so I indeed expect no problems once that is merged.

pllim commented:
@evertrol , I merged #5492. You should rebase this on top of master now, and see if tests pass.

astrofrog commented on CHANGES.rst:
```diff
@@ -180,6 +180,15 @@ API Changes
+    is used, an ``AstropyDeprecationWarning`` is issued. If both are
+    used, a ``TypeError`` is raised. [#5171]
+
+    This change affects the following functions and methods:
```
Can you make the changelog entry into a single paragraph?

astrofrog commented on CHANGES.rst:
```diff
@@ -222,6 +231,10 @@ API Changes
 
 - ``astropy.vo``
 
+  - ``VOSDatabase.to_json()`` now accepts an ``overwrite`` argument.
```
As above - if the argument has been renamed, make that clearer (at the moment it sounds like a new option has been added.

pllim commented:
It is approved, so merging. Thanks, @evertrol and @MSeifert04 !

p.s. "I am in fits and clobbered by trying to clobber a FITS file" does not sound that good by replacing "clobber" with "overwrite".

eteq commented:
👍 from me too (if a little late 😉 )
